source/Nevermore/Mapping/KeyAllocator.cs

```diff
@@ -28,15 +31,21 @@ public int NextId(string tableName)
             return allocation.Next();
         }
 
+        public async Task<int> NextIdAsync(string tableName, CancellationToken cancellationToken)
+        {
+            var allocation = allocations.GetOrAdd(tableName, _ => new Allocation(store, tableName, blockSize));
+            return await allocation.NextAsync(cancellationToken).ConfigureAwait(false);
+        }
+
         class Allocation
         {
             readonly IRelationalStore store;
             readonly string collectionName;
             readonly int blockSize;
-            readonly object sync = new object();
-            volatile int blockStart;
-            volatile int blockNext;
-            volatile int blockFinish;
+            readonly SemaphoreSlim sync = new(1, 1);
+            int blockStart;
+            int blockNext;
+            int blockFinish;
```

Comment:
The `volatile` keyword wasn't serving any purpose here. These were always protected by the `sync` mutex.